### ISSUE TITLE
Empty values can't be coerced

### ISSUE THREAD

**Issue by SleeplessByte** [hanami/model#273]
When a form is submitted and the value is empty (but present as key because it was in the form).

Coercing

```
'' Can't be coerced into Date
```

**Comment by jodosha** [comment]
@SleeplessByte Thanks for reporting this. We'll start reworking Lotus::Model low level mechanisms next week. So we'll take in account this problem.

For the time being, you can avoid to hit this problem, by validating your data.

**Comment by SleeplessByte** [comment]
I actually unset empty values before I pass them to the repository, because the `skip?` if blank? only skips it for validation. Not for creation.

**Comment by jodosha** [comment]
@SleeplessByte I'm confused. :cry: 

You should invoke methods on the repository, only if you're sure they are valid. This is a framework design. In pseudo code:

```ruby
if my_data_is_valid?
  MyRepository.create(my_entity_instance)
end
```

As opposite of other libraries like ActiveRecord, where the control happens inside the repository role:

```ruby
MyModel.create(my_data) # check data inside.
```

---

Do you have some code to share with us? It would help to understand better your use case :smile: Thanks!

**Comment by SleeplessByte** [comment]
@jodosha Sorry about the confusion.

Adding validations to the `action#params` actually solved this because it filters out empty `params`. This however is not clear from the documentation. After I updated my gems I think entity.valid? / action#params.valid? both ignored the empty values correctly.

**Comment by SleeplessByte** [comment]
Woops. The bug is still there off course for no validations.

**Comment by jodosha** [comment]
> Woops. The bug is still there off course for no validations.

Which version do you used before, and which one after the upgrade?

**Comment by SleeplessByte** [comment]
`0.5.0` to `0.6.x` + `lotusrb edge`.

I have not encountered the error since yesterday.

**Comment by jodosha** [comment]
@SleeplessByte Thank you. I suspect this could be related to https://github.com/lotus/validations/pull/82 /cc @hlegius

**Comment by hlegius** [comment]
@SleeplessByte if I got you correctly, you're trying to persist `params` directly thought your Entity, instead of use the validated params (a.k.a Lotus Validation output (with `.to_h` method) after you've checked if its valid (`.valid?`, `.invalid?`). This way, you'll get this coerce issue.

About blank values, since lotus/validations#82, you can have blank param's keys for both `size` and `type` validation' attributes.

If you still experiencing some unexpected behaviour on this, please, paste here some of your implementation. I'd be glad to help you!

**Comment by SleeplessByte** [comment]
@jodosha It was, that's why switching to edge resolved it.

@hlegius I was and I was not. I realised quickly that I had to do `params.valid?` to validate parameters. Right now I just have `Lotus::Validations` in both `model` and `params`. In the documentation `.to_h` is not mentioned, and it should if that is the way to go.

Right now the flow is something along the lines of:

````Ruby
params do
   param :foo do
       param :bar, baz: :validation
       ...
   end
end

def call( params )
  ...
  if params.valid?
     @my_model = MyModel.new( params[ :foo ] )
     @my_model = MyModelRepository.create( @my_model ) or raise SomeException
     ...
  end
  ...
end
````

And this works.

I had one problem using `presence: false (default)` and `format:` but that's resolved as well.

Long story short (\cc @jodosha ): Personally, I don't think it is clear enough that 
- one **must** validate `params` and not the model
- there is no safe handling of faulty values for model params
- using params after validation will sanitize empty values
- which validations are available by default, and which are after `skip?` (easy to check in the source, but still)
